src/rulesets/LinearAlgebra/factorization.jl

```diff
@@ -1,6 +1,207 @@
+)
+    F = lu(A, pivot; kwargs...)
+    function lu_pullback(ΔF::Composite)
+        Δfactors = ΔF.factors
+        Δfactors isa AbstractZero && return (NO_FIELDS, Δfactors, DoesNotExist())
+        factors = F.factors
+        ∂factors = eltype(A) <: Real ? real(Δfactors) : Δfactors
+        ∂A = similar(factors)
+        m, n = size(A)
+        q = min(m, n)
+        if m == n  # square A
+            # ∂A = P' * (L' \ (tril(L' * ∂L, -1) + triu(∂U * U')) / U')
+            L = UnitLowerTriangular(factors)
+            U = UpperTriangular(factors)
+            ∂U = UpperTriangular(∂factors)
+            ∂A = similar(∂factors)
+            tril!(copyto!(∂A, ∂factors), -1)
+            lmul!(L', ∂A)
+            copyto!(UpperTriangular(∂A), UpperTriangular(∂U * U'))
+            rdiv!(∂A, U')
+            ldiv!(L', ∂A)
+        elseif m < n  # wide A, system is [P*A1 P*A2] = [L*U1 L*U2]
+            ∂A = triu(∂factors)
+            @views begin
+                factors1 = factors[:, 1:q]
+                U2 = factors[:, (q + 1):end]
+                ∂A1 = ∂A[:, 1:q]
+                ∂A2 = ∂A[:, (q + 1):end]
+                ∂L = tril(∂factors[:, 1:q], -1)
+            end
+            L = UnitLowerTriangular(factors1)
+            U1 = UpperTriangular(factors1)
+            triu!(rmul!(∂A1, U1'))
+            ∂A1 .+= tril!(mul!(lmul!(L', ∂L), ∂A2, U2', -1, 1), -1)
+            rdiv!(∂A1, U1')
+            ldiv!(L', ∂A)
+        else  # tall A, system is [P1*A; P2*A] = [L1*U; L2*U]
+            ∂A = tril(∂factors, -1)
+            @views begin
+                factors1 = factors[1:q, :]
+                L2 = factors[(q + 1):end, :]
+                ∂A1 = ∂A[1:q, :]
+                ∂A2 = ∂A[(q + 1):end, :]
+                ∂U = triu(∂factors[1:q, :])
+            end
+            U = UpperTriangular(factors1)
+            L1 = UnitLowerTriangular(factors1)
+            tril!(lmul!(L1', ∂A1), -1)
+            ∂A1 .+= triu!(mul!(rmul!(∂U, U'), L2', ∂A2, -1, 1))
+            ldiv!(L1', ∂A1)
+            rdiv!(∂A, U')
+        end
+        if pivot === Val(true)
+            ∂A = ∂A[invperm(F.p), :]
+        end
+        return NO_FIELDS, ∂A, DoesNotExist()
+    end
+    return F, lu_pullback
+end
+
+#####
+##### functions of `LU`
+#####
+
+# this rrule is necessary because the primal mutates
+
+function rrule(::typeof(getproperty), F::TF, x::Symbol) where {T,TF<:LU{T,<:StridedMatrix{T}}}
+    function getproperty_LU_pullback(ΔY)
+        ∂factors = if x === :L
+            m, n = size(F.factors)
+            S = eltype(ΔY)
+            tril!([ΔY zeros(S, m, max(0, n - m))], -1)
+        elseif x === :U
+            m, n = size(F.factors)
+            S = eltype(ΔY)
+            triu!([ΔY; zeros(S, max(0, m - n), n)])
+        elseif x === :factors
+            Matrix(ΔY)
+        else
+            return (NO_FIELDS, DoesNotExist(), DoesNotExist())
+        end
+        ∂F = Composite{TF}(; factors=∂factors)
```

Comment:
why are `:L` and `:U` cases inside `:factors`, rather than e.g.
```suggestion
        ∂F = Composite{TF}(; L=∂L, U=∂U, factors=∂factors)
```
(with each `∂` thunked) ?

Reply:
We could do that, and originally I did something similar. Here are the pros and cons of this approach that I can think of:
Pros:
- don't do any work or allocate any matrices in the `rrule` for `getproperty`
- Accumulating derivatives is just simple addition of the `∂L`, `∂U`, and `∂factors`.

Cons:
- the `rrule` for `lu` now needs to do the work and encode the logic of combining `∂L`, `∂U`, and `∂factors`.
- it does not mirror the primal (`getproperty` generates `L` and `U`
- we can't test the rules using `ChainRulesTestUtils`

The last con ended up pushing me to reimplement everything to use `factors`. It's now simpler and more modular, plus easier to test. As an aside, there'd be no point in thunking here, since the cotangent `ΔY` should only be distributed to either `∂L`, `∂U`, or `∂factors`, and both the primal and pullback will by necessity be type-unstable.

Reply:
> it does not mirror the primal

I don't think that matters, does it? Except insofar as it leads to 

> we can't test the rules using ChainRulesTestUtils

Is not mirroring the primal what leads to this? Do we have an issue on why?
I'm wondering because this will apply to lots of Factorizations (LinearAlgebra has a habit of overloading getproperty to do a lot of work). Is this why SVD/Eigen/Cholesky tests don't use ChainRulesTestUtils?

Should we have some docs on when/why it might be better to mirror the primal? or is it something we could extend ChainRulesTestUtils to handle?

(Just trying to understand what's going on. Don't think any changes are needed in this PR so don't let this hold it up)

Reply:
> > it does not mirror the primal
> 
> I don't think that matters, does it? Except insofar as it leads to
> 
> > we can't test the rules using ChainRulesTestUtils

It doesn't matter much, except the work has to be done _some_time, which if not in the `rrule` for `getproperty` would be in the `rrule` for `lu`. But all things being equal, it then seems cleaner to mirror the actual primal.

> > we can't test the rules using ChainRulesTestUtils
> 
> Is not mirroring the primal what leads to this? Do we have an issue on why?
> I'm wondering because this will apply to lots of Factorizations (LinearAlgebra has a habit of overloading getproperty to do a lot of work). Is this why SVD/Eigen/Cholesky tests don't use ChainRulesTestUtils?
> 
> Should we have some docs on when/why it might be better to mirror the primal? or is it something we could extend ChainRulesTestUtils to handle?

I haven't thought about this enough to be very formal about it. `eigen(A)` for non-hermitian `StridedMatrix` `A` is blocked by instability of the phase convention for `eigen`, so it's a separate issue. For Hermitian `svd` and `eigen`, the problem is that FiniteDifferences and ChainRules only agree on `Hermitian` sensitvities when the function of interest is composed with a constructor, so it's also a separate issue. I'd need to look more closely at the `cholesky` and `svd` cases for `StridedMatrix`.